Add `SchemaSerializer.__reduce__` method to enable `pickle` serialization

edoakes opened the pull request:
<!-- Thank you for your contribution! -->

## Change Summary

Currently `pydantic` model definitions cannot be serialized using [cloudpickle](https://github.com/cloudpipe/cloudpickle). One of the underlying issues is that the `SchemaSerializer` object is not serializable (even using vanilla `pickle`).

This PR enables serialization for the `SchemaSerializer` by:

1. Holding a reference to the Python `schema` and `config` objects required to initialize the `SchemaSerializer`, and
2. defining a custom `__reduce__` method that instructs `cloudpickle` to reconstruct the `SchemaSerializer` from these objects.

I've also updated the existing `SchemaValidator` codepath to match this one, including fixing a bug where the `config` object was being lost on serialization.

## Related issue number

https://github.com/pydantic/pydantic/issues/6763

<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Documentation reflects the changes where applicable
* [ ] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

edoakes commented:
@davidhewitt I am unable to assign reviewers, but PTAL.

- I'm not very familiar with PyO3 or Rust in general, so please spot check that I'm not doing something nonsensical.
- Let me know what testing you'd like me to add.

edoakes commented on src/validators/url.rs:
```diff
@@ -498,7 +498,7 @@ fn check_sub_defaults(
             lib_url
                 .set_port(Some(default_port))
-                .map_err(|_| map_parse_err(ParseError::EmptyHost))?;
+                .map_err(|()| map_parse_err(ParseError::EmptyHost))?;
```
I'm not sure why, but the linter wanted me to change this...

adriangb replied:
You just have a newer version of rust (nightly?) than we're using in CI

edoakes replied:
It was the CI linter that was complaining... not sure how that happened since I didn't touch this file 🤷‍♂️

davidhewitt replied:
Possibly Rust 1.73 having dropped yesterday is related.

davidhewitt replied:
I just merged this separately in #1010

adriangb commented on tests/test.rs:
```diff
@@ -46,7 +46,7 @@ mod tests {
             }"#;
             let schema: &PyDict = py.eval(code, None, None).unwrap().extract().unwrap();
-            SchemaSerializer::py_new(schema, None).unwrap();
+            SchemaSerializer::py_new(py, schema, None).unwrap();
```
You can get rid of this argument I think

edoakes replied:
Compiler yelled at me to add this. I think If you call the Python constructor PyO3 passes it automatically but not when calling the method directly from rust?

adriangb commented on tests/test_garbage_collection.py:
```diff
@@ -27,7 +27,9 @@ class BaseModel:
 
         def __init_subclass__(cls) -> None:
-            cls.__schema__ = SchemaSerializer(core_schema.model_schema(cls, GC_TEST_SCHEMA_INNER))
+            cls.__schema__ = SchemaSerializer(
+                core_schema.model_schema(cls, GC_TEST_SCHEMA_INNER), config={'ser_json_timedelta': 'float'}
+            )
```
Curious why this was necessary?

edoakes replied:
Wasn't necessary, I just wanted to ensure that there was coverage of the garbage collection path when a config was passed given that I'm holding a reference to it now (existing tests didn't take a config).

adriangb commented:
You're pretty close! Please ignore the benchmarks and pydantic-integration test.

edoakes commented:
@adriangb oh wow, thanks for the pointer, it would've saved me a lot of trouble if I'd seen `SchemaValidator.__reduce__` 😓

Looks like I came to essentially the same solution. The differences are:

1. `SchemaValidator.__reduce__` is using different syntax to get the reference to the class and the Python schema argument. I don't understand the differences between these, let me know which I should opt for. To an uninformed eye, `SchemaSerializer::type_object(py).to_object(py)` looks less hacky than `let cls = slf.getattr("__class__")?;`.
2. I'm also saving and reconstructing the "config" object which `SchemaValidator.__reduce__` ignores. I'm not sure the impact of this.

Let me know if/how you'd like me to update these to match more closely.

davidhewitt reviewed (commented):
Generally looks great, thanks! Just posted my suggestion of how to implement `__reduce__`, and if you're happy with it then it'd be great to update `SchemaValidator` to work the same way. I also think it's probably a bug if we're not pickling the config for `SchemaValidator`.

davidhewitt commented on src/serializers/mod.rs:
```diff
@@ -174,6 +183,14 @@ impl SchemaSerializer {
         Ok(py_bytes.into())
     }
 
+    pub fn __reduce__(&self, py: Python) -> PyResult<(PyObject, (PyObject, PyObject))> {
+        // Enables support for `cloudpickle` serialization.
+        Ok((
+            SchemaSerializer::type_object(py).to_object(py),
+            (self.py_schema.to_object(py), self.py_config.to_object(py)),
+        ))
+    }
```
Here's my take:

```suggestion
    pub fn __reduce__(slf: &PyCell<Self>) -> PyResult<(&PyType, (PyObject, PyObject))> {
        let py = slf.py();
        let args = (slf.get().schema.to_object(py), slf.get().py_config.to_object(py));
        let cls = slf.get_type();
        Ok((cls, args))
    }
```

You'll need to update the `#[pyclass]` definition above for `SchemaSerializer` to include `#[pyclass(frozen)]`.

Justifications:

1. `slf: &PyCell<Self>` is equivalent to `self` Python object
2. `slf.get()` gives read access to the Rust data as long as it's `frozen`
3. I think we want to look up `type(self)` rather than hard-code `SchemaSerializer` type in case we allow subclassing later. `slf.get_type()` is a cleaner way to do it than trying to use `getattr`.

edoakes replied:
Makes sense, thanks for the pointers, I'll update!

edoakes commented:
Ok @davidhewitt I've updated the PR to include the suggested changes.

- Now using `slf.get_type()` to get the reference to the object type in `__reduce__`.
- Updated `SchemaValidator.__reduce__` to match and to fix the bug with the config not being serialized.
- I also updated some naming to make them consistent, please let me know if I've taken too much liberty there.
- Added test cases as `test_pickling.py` for both the serializer and validator.

edoakes commented on tests/validators/test_datetime.py:
```diff
@@ -480,17 +479,6 @@ def test_tz_constraint_wrong():
         validate_core_schema(core_schema.datetime_schema(tz_constraint='wrong'))
 
 
-def test_tz_pickle() -> None:
```
This is moved to `test_pickling.py`

edoakes commented on tests/validators/test_pickling.py:
```diff
@@ -0,0 +1,55 @@
+        v.validate_test([])
+
+
+def test_schema_validator_containing_config():
```
I verified that this newly-added test case fails without my changes to `SchemaValidator.__reduce__` (config object is lost).

davidhewitt reviewed (approved):
Looks great to me, many thanks 👍
